### ISSUE TITLE
[LifetimeSafety] False-negative for dangling std::optional temporary

### ISSUE THREAD

**Issue by usx95** [llvm/llvm-project#175893]
```cpp
#include <optional>
#include <vector>

const int& VectorTemp() {
    return std::vector<int>().at(0);  // Error. Good.
}

const int& OptionalTemp() {
    return std::optional<int>().value();  // No error. Bad.
}

const int& OptionalLocal() {
    auto local = std::optional<int>();
    return local.value();  // Error. Good.
}
```
https://godbolt.org/z/s18e48K1f
